heal/vlmd/extract/utils.py

```diff
@@ -0,0 +1,293 @@
+    # flatten if type object with properties
+    items = []
+    for key, value in dictionary.items():
+        new_key = str(parent_key) + sep + key if parent_key else key
+        prop = schema["properties"].get(key, {})
+        childprops = prop.get("properties")
+        childitem_props = prop.get("items", {}).get("properties")
+
+        if childitem_props:
+            for i, _value in enumerate(value):
+                item = flatten_to_jsonpath(
+                    dictionary=_value,
+                    schema=prop,
+                    parent_key=new_key,
+                    sep=f"[{str(i)}]{sep}",
+                )
+                items.extend(item.items())
+        elif childprops:
+            item = flatten_to_jsonpath(
+                dictionary=value, schema=prop, parent_key=new_key, sep=sep
+            )
+            items.extend(item.items())
+
+        else:
+            items.append((new_key, value))
+
+    return dict(items)
+
+
+def unflatten_from_jsonpath(field):
+    """
+    Converts a flattened dictionary with key names conforming to
+    JSONpath notation to the nested dictionary format.
+    """
+    field_json = {}
+
+    for prop_path, prop in field.items():
+        prop_json = field_json
+
+        nested_names = prop_path.split(".")
+        for prop_name in nested_names[:-1]:
+            if "[" in prop_name:
+                array_name, array_index = prop_name.split("[")
+                array_index = int(array_index[:-1])
+                if array_name not in prop_json:
+                    prop_json[array_name] = [{} for _ in range(array_index + 1)]
+                elif len(prop_json[array_name]) <= array_index:
+                    prop_json[array_name].extend(
+                        [
+                            {}
+                            for _ in range(array_index - len(prop_json[array_name]) + 1)
+                        ]
+                    )
+                prop_json = prop_json[array_name][array_index]
+            else:
+                if prop_name not in prop_json:
+                    prop_json[prop_name] = {}
+                prop_json = prop_json[prop_name]
+
+        last_prop_name = nested_names[-1]
+        if "[" in last_prop_name:
+            array_name, array_index = last_prop_name.split("[")
+            array_index = int(array_index[:-1])
+            if array_name not in prop_json:
+                prop_json[array_name] = [{} for _ in range(array_index + 1)]
+            elif len(prop_json[array_name]) <= array_index:
+                prop_json[array_name].extend(
+                    [{} for _ in range(array_index - len(prop_json[array_name]) + 1)]
+                )
+            if isinstance(prop_json[array_name][array_index], dict):
+                prop_json[array_name][array_index].update({array_name: prop})
+            else:
+                prop_json[array_name][array_index] = {array_name: prop}
+        else:
+            prop_json[last_prop_name] = prop
```

Comment:
L195-L210 is very similar to L176-L193 above, except it does something special at the end for the last element. Probably better if written this way to combine the two
```
nested_names_length = len(nested_names)
for i, prop_name in enumerate(nested_names):
  if "[" in prop_name:
  ...
  if i == nested_names_length - 1
  // do something for the last element
  else:
  // do other things for other elements
```

Reply:
These blocks have been combined as described in your comments.